cmd/controller/controller.go

```diff
@@ -0,0 +1,163 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"errors"
+	"fmt"
+	"os"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/golang/glog"
+	apis "github.com/openebs/node-disk-manager/pkg/apis/ndm/v1alpha1"
+	clientset "github.com/openebs/node-disk-manager/pkg/client/clientset/versioned"
+	"github.com/openebs/node-disk-manager/pkg/signals"
+	"github.com/openebs/node-disk-manager/pkg/util"
+)
+
+const (
+	NDMKind     = "Disk"                   // NDMKind is the CR kind.
+	NDMVersion  = "openebs.io/v1alpha1"    // NDMVersion is the CR version.
+	NDMHostKey  = "kubernetes.io/hostname" // NDMHostKey is host name label prefix.
+	NDMActive   = "Active"                 // NDMActive is constant for active resource status.
+	NDMInactive = "Inactive"               // NDMInactive is constant for inactive resource status.
+)
+
+// ProbeChannel is used to send a copy of controller object to each probe.
+// Each probe can get the copy of controller struct any time they
+// need to read the channel.
+var ProbeChannel = make(chan *Controller)
+
+// Controller is the controller implementation for do resources
+type Controller struct {
+	HostName      string               // host name is host name in which disk is attached
+	KubeClientset kubernetes.Interface // kubeclientset is standard kubernetes clientset
+	Clientset     clientset.Interface  // sampleclientset is clientset for our own API group
+}
+
+// newController returns a controller pointer for any error case it will return nil
+func newController(kuberconfig string) (*Controller, error) {
+	masterURL := ""
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		if kuberconfig == "" {
+			return nil, errors.New("kubeconfig is empty")
+		}
+		cfg, err = clientcmd.BuildConfigFromFlags(masterURL, kuberconfig)
+		if err != nil {
+			return nil, err
+		}
+	}
+	kubeClient, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	crdClient, err := clientset.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+	host, ret := os.LookupEnv("NODE_NAME")
+	if ret != true {
+		return nil, errors.New("error building hostname")
+	}
+	controller := &Controller{
+		HostName:      host,
+		KubeClientset: kubeClient,
+		Clientset:     crdClient,
+	}
+	return controller, nil
+}
+
+// Start scans the locally attached disks and push them to etcd.
+// This function is called when we execute cli command ndm start.
+func Start(kuberconfig string) {
+	// set up signals so we handle the first shutdown signal gracefully
+	stopCh := signals.SetupSignalHandler()
+	controller, err := newController(kuberconfig)
+	if err != nil {
+		glog.Fatal(err)
+	}
+	if err = controller.run(2, stopCh); err != nil {
+		glog.Fatalf("error running controller: %s", err.Error())
+	}
+}
+
+// run will discover the local disks and push them to the etcd.
```

Comment:
It appears that this comment needs to be updated with the refactored code.

Reply:
Updated comment added.